File: autojs/src/main/java/com/stardust/autojs/engine/module/AssetAndUrlModuleSourceProvider.java
package com.stardust.autojs.engine.module;

import android.content.res.AssetManager;

import com.stardust.autojs.engine.encryption.ScriptEncryption;
import com.stardust.autojs.script.EncryptedScriptFileHeader;

import org.mozilla.javascript.commonjs.module.provider.ModuleSource;

import java.io.ByteArrayInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLConnection;
import java.util.List;

/**
 * Created by Stardust on 2017/5/9.
 */

public class AssetAndUrlModuleSourceProvider extends UrlModuleSourceProvider {

    private android.content.Context mContext;
    private final URI mBaseURI;
    private final String mAssetDirPath;
    private final AssetManager mAssetManager;

    public AssetAndUrlModuleSourceProvider(android.content.Context context, String assetDirPath, List<URI> list) {
        super(list, null);
        mContext = context;
        mAssetDirPath = assetDirPath;
        mBaseURI = URI.create("file:///android_asset/" + assetDirPath);
        mAssetManager = mContext.getAssets();
    }

    @Override
    protected ModuleSource loadFromPrivilegedLocations(String moduleId, Object validator) throws IOException, URISyntaxException {
        String moduleIdWithExtension = moduleId;
        if (!moduleIdWithExtension.endsWith(".js")) {
            moduleIdWithExtension += ".js";
        }
        try {
            return new ModuleSource(new InputStreamReader(mAssetManager.open(mAssetDirPath + "/" + moduleIdWithExtension)), null,
                    new URI(mBaseURI.toString() + "/" + moduleIdWithExtension), mBaseURI, validator);
        } catch (FileNotFoundException e) {
            return super.loadFromPrivilegedLocations(moduleId, validator);
        }
    }

    @Override
    protected ModuleSource loadFromUri(URI uri, URI base, Object validator) throws IOException, URISyntaxException {
        ModuleSource moduleSource = super.loadFromUri(uri, base, validator);
        if (moduleSource == null && uri.getPath().startsWith("/android_asset/")) {
            return new ModuleSource(new InputStreamReader(mAssetManager.open(uri.getPath().replace("/android_asset/", ""))), null,
                    uri, mBaseURI, validator);
        }
        return moduleSource;
    }

    @Override
    protected Reader getReader(URLConnection urlConnection) throws IOException {
        InputStream stream = urlConnection.getInputStream();
        byte[] bytes = new byte[stream.available()];
        stream.read(bytes);
        stream.close();
        if (EncryptedScriptFileHeader.INSTANCE.isValidFile(bytes)) {
            byte[] clearText = ScriptEncryption.INSTANCE.decrypt(bytes, EncryptedScriptFileHeader.BLOCK_SIZE, bytes.length);
            return new InputStreamReader(new ByteArrayInputStream(clearText));
        }
        return new InputStreamReader(new ByteArrayInputStream(bytes));
    }
}

File: autojs/src/main/java/com/stardust/autojs/engine/module/AssetAndUrlModuleSourceProvider.kt
package com.stardust.autojs.engine.module

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import com.google.gson.Gson
import okhttp3.OkHttpClient
import okhttp3.Request
import org.mozilla.javascript.commonjs.module.provider.ModuleSource
import org.mozilla.javascript.commonjs.module.provider.ModuleSourceProviderBase
import java.io.File
import java.io.InputStream
import java.io.InputStreamReader
import java.net.URI

class AssetAndUrlModuleSourceProvider(
    context: Context,
    assetDirPath: String,
    list: List<URI>? = null
) : ModuleSourceProviderBase() {
    val mContext = context
    private val okHttpClient = OkHttpClient.Builder().followRedirects(true).build()
    private val contentResolver: ContentResolver = context.contentResolver
    private val moduleSources: ArrayList<URI> = arrayListOf(mBaseURI, npmModuleSource)

    companion object {
        val mBaseURI: URI = URI.create("file:/android_asset/modules")
        val npmModuleSource:URI = URI.create("file:/android_asset/modules/npm")
    }
    //初始化脚本以及启动文件只会从此方法加载模块，子模块加载没有以"./"或"../"开头的模块也会从此方法加载
    override fun loadFromPrivilegedLocations(moduleId: String, validator: Any?): ModuleSource? {
        //println("加载私有模块：$moduleId")
        val uri = if (moduleId.startsWith("/")) {
            File(moduleId).toURI()
        } else if (moduleId.startsWith("http://") || moduleId.startsWith("https://")) {
            URI.create(moduleId)
        } else null
        if (uri != null) {
            return loadFromUri(uri, File(uri.path).parentFile?.toURI(), validator)
        }
        for (baseUri in moduleSources) {
            val sourceUri = URI.create("$baseUri/$moduleId")
            val moduleSource = loadFromUri(sourceUri, baseUri, validator)
            if (moduleSource != null) {
                return moduleSource
            }
        }
        return null
    }

    //这里处理node_module目录的模块
    override fun loadFromFallbackLocations(moduleId: String, validator: Any?): ModuleSource? {
        return super.loadFromFallbackLocations(moduleId, validator)
    }

    //子模块以相对路径加载时调用此方法
    override fun loadFromUri(uri: URI, base: URI?, validator: Any?): ModuleSource? {
        var uri = uri
        if (uri.scheme == null) uri = File(uri.path).toURI()
        //println("加载模块：$uri")
        if (uri.scheme == "http" || uri.scheme == "https") {
            return loadFromHttp(uri, base, validator)
        }
        val moduleSource = loadAt(uri, base, validator) ?: loadAt(
            File(uri.path + ".js").toURI(), base, validator
        )
        if (moduleSource != null) {
            return moduleSource
        }
        //尝试从目录加载
        //尝试读取package.json指定的文件
        val mainFile: URI? = try {
            val packageFile = File(uri.path, "package.json")
            val json = Gson().fromJson<Map<String, Any>>(
                InputStreamReader(packageFile.inputStream()),
                Map::class.java
            )
            val main = json["main"] as String?
            main?.let {
                packageFile.toURI().resolve(main)
            }
        } catch (e: Exception) {
            null
        }
        val main: URI = mainFile ?: File(uri.path, "index.js").toURI()
        return loadAt(main, uri, validator)
    }

    private fun loadAt(uri: URI, base: URI?, validator: Any?): ModuleSource? {
        if (uri.scheme == "http" || uri.scheme == "https") {
            return loadFromHttp(uri, base, validator)
        }
        return try {
            val inputStream = if (uri.path.startsWith("/android_asset/")) {
                mContext.assets.open(uri.path.replace("/android_asset/", ""))
            } else contentResolver.openInputStream(Uri.parse(uri.toString()))
            if (inputStream != null) {
                createModuleSource(inputStream, uri, base, validator)
            } else null
        } catch (e: Exception) {
            null
        }
    }

    private fun loadFromHttp(uri: URI, base: URI?, validator: Any?): ModuleSource? {
        return try {
            Request.Builder().url(uri.toString()).build().let { request ->
                val response = okHttpClient.newCall(request).execute()
                if (!response.isSuccessful) {
                    response.close()
                    return null
                }
                response.body?.let {
                    val charset = it.contentType()?.charset()?.toString() ?: "utf-8"
                    return createModuleSource(it.byteStream(), uri, base, validator, charset)
                }
            }
        } catch (e: Exception) {
            null
        }
    }

    private fun createModuleSource(
        inputStream: InputStream,
        uri: URI,
        base: URI?,
        validator: Any?,
        charset: String? = null
    ): ModuleSource {
        val id = if (uri.scheme == "file") {
            URI.create(uri.path)
        } else uri
        return ModuleSource(
            InputStreamReader(inputStream, charset ?: "utf-8"),
            null,
            id,
            base,
            validator
        )
    }
}